### PR TITLE
tickets/DM-8457: Wrap with pybind11

### PR DESCRIPTION
**Pull request by fred3m**

### CODE REVIEW

File: python/lsst/meas/deblender/SConscript [hunk 1]
```diff
@@ -1,4 +1,3 @@
-# -*- python -*-
+## -*- python -*-
```
Comment (pschella) [inline python/lsst/meas/deblender/SConscript]: remove extra comment

File: python/lsst/meas/deblender/baseline.cc [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+using namespace pybind11::literals;
+
+namespace lsst {
+namespace meas {
+namespace deblender {
+
+namespace{
+    template <typename ImagePixelT,
+              typename MaskPixelT=lsst::afw::image::MaskPixel,
+              typename VariancePixelT=lsst::afw::image::VariancePixel>
```
Comment (pschella) [inline python/lsst/meas/deblender/baseline.cc]: Note that in previous cases we have just used the defaults and only kept `PixelT`.

File: python/lsst/meas/deblender/baseline.cc [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    void declareBaselineUtils(py::module & mod, const std::string & suffix){
+        typedef typename lsst::afw::image::MaskedImage<ImagePixelT, MaskPixelT, VariancePixelT> MaskedImageT;
+        typedef typename PTR(lsst::afw::image::Image<ImagePixelT>) ImagePtrT;
+        typedef typename PTR(lsst::afw::detection::Footprint) FootprintPtrT;
```
Comment (pschella) [inline python/lsst/meas/deblender/baseline.cc]: I prefer `using`.

Comment (pschella) [inline python/lsst/meas/deblender/baseline.cc]: But it doesn't matter.

Comment (pschella) [inline python/lsst/meas/deblender/baseline.cc]: And you don't need the `typename`.

File: python/lsst/meas/deblender/baseline.cc [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                          ImagePtrT templ_sum,
+                          std::vector<bool> const& ispsf,
+                          std::vector<int>  const& pkx,
+                          std::vector<int>  const& pky,
+                          int strayFluxOptions,
+                          double clipStrayFluxFraction
+                     ){
+            typedef typename std::vector<std::shared_ptr<typename lsst::afw::detection::HeavyFootprint<ImagePixelT,MaskPixelT,VariancePixelT> > > HeavyFootprintPtrList;
+            
+            std::vector<typename PTR(lsst::afw::image::MaskedImage<ImagePixelT, MaskPixelT, VariancePixelT>)> result;
```
Comment (pschella) [inline python/lsst/meas/deblender/baseline.cc]: `typename` not needed

File: python/lsst/meas/deblender/baseline.cc [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                          std::vector<typename PTR(lsst::afw::image::Image<ImagePixelT>)> templates,
+                          std::vector<std::shared_ptr<lsst::afw::detection::Footprint> > templ_footprints,
+                          ImagePtrT templ_sum,
+                          std::vector<bool> const& ispsf,
+                          std::vector<int>  const& pkx,
+                          std::vector<int>  const& pky,
+                          int strayFluxOptions,
+                          double clipStrayFluxFraction
+                     ){
+            typedef typename std::vector<std::shared_ptr<typename lsst::afw::detection::HeavyFootprint<ImagePixelT,MaskPixelT,VariancePixelT> > > HeavyFootprintPtrList;
```
Comment (pschella) [inline python/lsst/meas/deblender/baseline.cc]: `typename` not needed

File: python/lsst/meas/deblender/baseline.cc [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        });
+        cls.def_static("medianFilter", &BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::medianFilter);
+        cls.def_static("makeMonotonic",
+                       &BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::makeMonotonic);
+        // The function has a std::pair return value and a referenced HeavyFootprint list (strays) that is
+        // updated, which must be returned to avoid using OPAQUE types.
+        cls.def_static("apportionFlux",
+                       [](MaskedImageT const& img,
+                          lsst::afw::detection::Footprint const& foot,
+                          std::vector<typename PTR(lsst::afw::image::Image<ImagePixelT>)> templates,
```
Comment (pschella) [inline python/lsst/meas/deblender/baseline.cc]: `typename` not needed

File: python/lsst/meas/deblender/baseline.cc [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            
+            result = BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::buildSymmetricTemplate(
+                img, foot, pk, sigma1, minZero, patchEdges, &patchedEdges);
+            return py::make_tuple(result.first, result.second, patchedEdges);
+        });
+        cls.def_static("medianFilter", &BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::medianFilter);
+        cls.def_static("makeMonotonic",
+                       &BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::makeMonotonic);
+        // The function has a std::pair return value and a referenced HeavyFootprint list (strays) that is
+        // updated, which must be returned to avoid using OPAQUE types.
```
Comment (pschella) [inline python/lsst/meas/deblender/baseline.cc]: This comment is not clear to me. What exactly is it avoiding and how?

File: python/lsst/meas/deblender/baseline.cc (hunk 3 above)
Comment (fred3m) [inline python/lsst/meas/deblender/baseline.cc]: These lines are just copied directly from the header files. I thought for continuity I would keep the typedefs exactly the same. I could eliminate the `typename`s if you like but I would prefer to leave them.

File: python/lsst/meas/deblender/baseline.py [hunk 8]
```diff
@@ -131,9 +134,11 @@ def __init__(self, footprints, maskedImages, psfs, psffwhms, log, filters=None,
             self.deblendedParents[self.filters[n]] = dp
 
         # Group the peaks in each color
+        self.peaks = []
```
Comment (pschella) [inline python/lsst/meas/deblender/baseline.py]: Why is this needed?

File: tests/testEdges.py [hunk 9]
```diff
@@ -367,8 +366,6 @@ def myimshow(*args, **kwargs):
             print('Wrote', fn)
 
 
-#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
```
Comment (pschella) [inline tests/testEdges.py]: Are we sure this is new standard?

Comment (fred3m) [inline tests/testEdges.py]: Yes, these should have been stripped out during the py.test port. This one must have been missed.

File: python/lsst/meas/deblender/baseline.py (hunk 8 above)
Comment (fred3m) [inline python/lsst/meas/deblender/baseline.py]: This commit was made on the main stack, but had not been merged yet. So I included it here so that it is one last conflict when we merge pybind11 with master (it is the same exact commit, I created this branch from the one I modified but have not merged yet).

File: python/lsst/meas/deblender/baseline.cc (hunk 7 above)
Comment (fred3m) [inline python/lsst/meas/deblender/baseline.cc]: The problem is the c++ function takes a pointer to an empty vector as an input and adds elements to the vector. But when a python list is passed to the function the list is not populated. So instead we omit the vector/list as an input, create it in the lambda, and add the modified vector to the returns.

Comment (fred3m) [inline python/lsst/meas/deblender/baseline.cc]: I updated the comment to be more clear
